### PR TITLE
Use `Self` rather than concrete types, remove `cast`s

### PR DESCRIPTION
**Pull request by max-sixty**
This should also allow for subtyping

### CODE REVIEW

File: xarray/core/common.py [hunk 1]
```diff
@@ -381,11 +382,11 @@ class DataWithCoords(AttrAccessMixin):
     __slots__ = ("_close",)
 
     def squeeze(
-        self: T_DataWithCoords,
+        self,
         dim: Hashable | Iterable[Hashable] | None = None,
         drop: bool = False,
         axis: int | Iterable[int] | None = None,
-    ) -> T_DataWithCoords:
+    ) -> Self:
```
Comment (Illviljan) [inline xarray/core/common.py:389]: Something I've been wondering about with Self usage: Is a copied version of self ok? 

`squeeze` uses `isel` which in the end creates a new copy with `types(self)(dims=...,  data=...)`

Comment (max-sixty) [reply xarray/core/common.py:389]: IIUC, it means that it's the _type_ of `self`, not that it's literally the object. So a copy would be completely fine, given it retains its type. (or tell me if I'm misunderstanding)

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -2809,7 +2819,7 @@ def unstack(
         dim: Dims = None,
         fill_value: Any = dtypes.NA,
         sparse: bool = False,
-    ) -> DataArray:
+    ) -> Self:
```
Comment (headtr1ck) [inline xarray/core/dataarray.py:2822]: The comment above this method is now obsolete. (There should be more of these cases, just look for the mypy issue in the code)

File: xarray/core/dataarray.py [hunk 3]
```diff
@@ -4414,13 +4420,13 @@ def to_iris(self) -> iris_Cube:
...
     @classmethod
-    def from_iris(cls, cube: iris_Cube) -> DataArray:
+    def from_iris(cls, cube: iris_Cube) -> Self:
         """Convert a iris.cube.Cube into an xarray.DataArray"""
         from xarray.convert import from_iris
 
         return from_iris(cube)
 
-    def _all_compat(self: T_DataArray, other: T_DataArray, compat_str: str) -> bool:
+    def _all_compat(self, other: T_DataArray, compat_str: str) -> bool:
```
Comment (headtr1ck) [inline xarray/core/dataarray.py:4429]: Here I'm not sure what the correct type is. A TypeVar in a method definition without using this TypeVar in the return type is useless.

The question is if other should be DataArray or Self?

Same with the next few changes below

File: xarray/core/dataarray.py [hunk 4]
```diff
@@ -4661,7 +4667,7 @@ def _binary_op(
 
         return self._replace(variable, coords, name, indexes=indexes)
 
-    def _inplace_binary_op(self: T_DataArray, other: Any, f: Callable) -> T_DataArray:
+    def _inplace_binary_op(self, other: Any, f: Callable) -> Self:
```
Comment (headtr1ck) [inline xarray/core/dataarray.py:4670]:
```suggestion
    def _inplace_binary_op(self, other: DaCompatible, f: Callable) -> Self:
```

File: xarray/core/dataarray.py [hunk 5]
```diff
@@ -4892,10 +4898,10 @@ def imag(self: T_DataArray) -> T_DataArray:
         return self._replace(self.variable.imag)
 
     def dot(
-        self: T_DataArray,
+        self,
         other: T_DataArray,
         dims: Dims = None,
-    ) -> T_DataArray:
+    ) -> Self:
```
Comment (headtr1ck) [inline xarray/core/dataarray.py:4904]: Again other needs to be changed

File: xarray/core/dataarray.py [hunk 6]
```diff
@@ -5908,7 +5914,7 @@ def argmin(
         axis: int | None = None,
         keep_attrs: bool | None = None,
         skipna: bool | None = None,
-    ) -> DataArray | dict[Hashable, DataArray]:
+    ) -> Self | dict[Hashable, DataArray]:
```
Comment (headtr1ck) [inline xarray/core/dataarray.py:5917]:
```suggestion
    ) -> Self | dict[Hashable, Self]:
```

File: xarray/core/dataarray.py [hunk 7]
```diff
@@ -6010,7 +6016,7 @@ def argmax(
...
```
Comment (headtr1ck) [inline xarray/core/dataarray.py:6019]:
```suggestion
    ) -> Self | dict[Hashable, Self]:
```

File: xarray/core/dataset.py [hunk 8]
```diff
@@ -3331,20 +3331,22 @@ def broadcast_like(
 
         dims_map, common_coords = _get_broadcast_dims_map_common_coords(args, exclude)
 
+        # Currently requires a `cast` and returns `Dataset` rather than `Self`, haven't
+        # worked through exactly why
         return _broadcast_helper(
-            cast("T_Dataset", args[1]), exclude, dims_map, common_coords
+            cast("Dataset", args[1]), exclude, dims_map, common_coords
```
Comment (headtr1ck) [inline xarray/core/dataset.py:3337]: I guess we have to look at what align does to subclasses

**Review by headtr1ck — COMMENTED** [review]
That's something I've wanted to do for quite a while now. Thanks!

File: xarray/core/common.py (hunk 1 above)
Comment (Illviljan) [reply xarray/core/common.py:389]: Nice, then there's a lot of places that could use Self!

File: xarray/core/dataarray.py (hunk 2 above)
Comment (max-sixty) [reply xarray/core/dataarray.py:2822]: Great, lots of hits for `12846`!

File: xarray/core/dataarray.py (hunk 5 above)
Comment (max-sixty) [reply xarray/core/dataarray.py:4904]: Good spot — found a few of these. I suspect I haven't quite got them all though.

V happy to add a few more if you spot them. I suspect the only way to catch em all is with tests though...

File: xarray/core/dataset.py (hunk 8 above)
Comment (max-sixty) [reply xarray/core/dataset.py:3337]: Yes, I looked at this, but eventually thought it wasn't worth gating on — it's quite complicated. It's something like "the 'biggest' class" — so "da + da = da", but "da + ds = ds"...

File: xarray/core/dataarray.py (hunk 3 above)
Comment (max-sixty) [reply xarray/core/dataarray.py:4429]: > A TypeVar in a method definition without using this TypeVar in the return type is useless.

Isn't it at least restricting it to `DataArray` and its subtypes? Or you mean useless relative to that?

File: xarray/core/dataset.py (hunk 8 above)
Comment (max-sixty) [reply xarray/core/dataset.py]: I adjusted these a bit, see what you think. Allowed removing some `cast`s

File: xarray/core/dataarray.py (hunk 3 above)
Comment (headtr1ck) [reply xarray/core/dataarray.py]: You can always assign a subclass to an argument.

Comment (max-sixty) [reply xarray/core/dataarray.py]: But `T_DataArray` is useless relative to having nothing there? Or it's useless relative to just having `DataArray`?

Comment (headtr1ck) [reply xarray/core/dataarray.py]: > But `T_DataArray` is useless relative to having nothing there? Or it's useless relative to just having `DataArray`?

Ah ok, it is useless relative to just `DataArray`

File: xarray/core/coordinates.py [hunk 9]
```diff
@@ -664,6 +664,8 @@ def copy(
         variables = {
             k: v._copy(deep=deep, memo=memo) for k, v in self.variables.items()
         }
+        # Currently this requires using the concrete type of `Coordinates`, rather than
+        # `Self`, ref #8216
```
Comment (Illviljan) [inline xarray/core/coordinates.py:668]: why doesn't type(self) work here?
```python
        return type(self)._construct_direct(
            coords=variables, indexes=dict(self.xindexes), dims=dict(self.sizes)
        )
```

Comment (max-sixty) [reply xarray/core/coordinates.py]: I went to answer this question, and then got it working :)

Comment (max-sixty) [reply xarray/core/coordinates.py]: OK actually the typing worked but the tests didn't; switching back. Added a TODO but won't chase it down atm...

**Pull request merged**